src/phase0/state_transition/slot_processing.rs

```diff
@@ -67,11 +75,137 @@ pub fn process_slot<
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+    const MAX_PROPOSER_SLASHINGS: usize,
+    const MAX_ATTESTER_SLASHINGS: usize,
+    const MAX_ATTESTATIONS: usize,
+    const MAX_DEPOSITS: usize,
+    const MAX_VOLUNTARY_EXITS: usize,
+>(
+    state: &mut BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    signed_block: &SignedBeaconBlock<
+        MAX_PROPOSER_SLASHINGS,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        MAX_ATTESTER_SLASHINGS,
+        MAX_ATTESTATIONS,
+        MAX_DEPOSITS,
+        MAX_VOLUNTARY_EXITS,
+    >,
+    validate_result: Option<bool>,
+    context: &Context,
+) -> Result<(), Error> {
+    let merkleization_context = context.for_merkleization();
+    let validate_result = validate_result.unwrap_or(true);
+    let block = &signed_block.message;
+
+    process_slots(state, block.slot, context)?;
+    if validate_result {
+        verify_block_signature(state, signed_block, context)?;
+    }
+    process_block(state, block, context)?;
+    if validate_result && block.state_root != state.hash_tree_root(merkleization_context)? {
+        return Err(InvalidStateRoot);
+    }
+
+    Ok(())
+}
+
+pub fn get_matching_source_attestations<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    epoch: Epoch,
+    context: &Context,
+) -> Result<Vec<PendingAttestation<MAX_VALIDATORS_PER_COMMITTEE>>, Error> {
+    let previous_epoch = get_previous_epoch(state, context);
+    let current_epoch = get_current_epoch(state, context);
+
+    if previous_epoch != epoch && current_epoch != epoch {
+        return Err(InvalidEpoch {
+            previous: previous_epoch,
+            current: current_epoch,
+        });
+    }
+
+    if epoch == current_epoch {
+        return Ok(state.current_epoch_attestations.to_vec());
+    }
+    Ok(state.previous_epoch_attestations.to_vec())
```

Comment:
we should be able to just return a ref to the correct attestation list, rather than allocate an entirely new Vec for them...

i'd suggest something like this: https://gist.github.com/rust-play/bda5e26dcd644ec30b358966eb382731